src/machine/machine_stm32l0x2.go

```diff
@@ -0,0 +1,108 @@
+// +build stm32l0x2
+
+package machine
+
+// Peripheral abstraction layer for the stm32l0x2
+
+import (
+	"device/stm32"
+	"runtime/interrupt"
+)
+
+func CPUFrequency() uint32 {
+	return 32000000
+}
+
+//---------- UART related types and code
+
+// UART representation
+type UART struct {
+	Buffer          *RingBuffer
+	Bus             *stm32.USART_Type
+	Interrupt       interrupt.Interrupt
+	AltFuncSelector stm32.AltFunc
+}
+
+// Configure the UART.
+func (uart UART) configurePins(config UARTConfig) {
+	// enable the alternate functions on the TX and RX pins
+	config.TX.ConfigureAltFunc(PinConfig{Mode: PinModeUARTTX}, uart.AltFuncSelector)
+	config.RX.ConfigureAltFunc(PinConfig{Mode: PinModeUARTRX}, uart.AltFuncSelector)
+}
+
+// UART baudrate calc based on the bus and clockspeed
+func (uart UART) getBaudRateDivisor(baudRate uint32) uint32 {
+	var clock, rate uint32
+	switch uart.Bus {
+	case stm32.LPUSART1:
+		clock = CPUFrequency() / 2 // APB1 Frequency
+		rate = uint32((256 * clock) / baudRate)
+	case stm32.USART1:
+		clock = CPUFrequency() / 2 // APB2 Frequency
+		rate = uint32(clock / baudRate)
+	case stm32.USART2:
+		clock = CPUFrequency() / 2 // APB1 Frequency
+		rate = uint32(clock / baudRate)
+	}
+
+	return rate
+}
+
+//---------- SPI related types and code
+
+// SPI on the STM32Fxxx using MODER / alternate function pins
+type SPI struct {
+	Bus             *stm32.SPI_Type
+	AltFuncSelector stm32.AltFunc
+}
+
+// Set baud rate for SPI
+func (spi SPI) getBaudRate(config SPIConfig) uint32 {
+	var conf uint32
+
+	localFrequency := config.Frequency
+	if spi.Bus != stm32.SPI1 {
+		// Assume it's SPI2 or SPI3 on APB1 at 1/2 the clock frequency of APB2, so
+		//  we want to pretend to request 2x the baudrate asked for
+		localFrequency = localFrequency * 2
+	}
+
+	// set frequency dependent on PCLK prescaler. Since these are rather weird
+	// speeds due to the CPU freqency, pick a range up to that frquency for
+	// clients to use more human-understandable numbers, e.g. nearest 100KHz
+
+	// These are based on APB2 clock frquency (84MHz on the discovery board)
+	// TODO: also include the MCU/APB clock setting in the equation
+	switch true {
+	case localFrequency < 328125:
+		conf = stm32.SPI_PCLK_256
```

Comment:
It would be useful to make this calculation explicit. For example:

```go
const apb2Frequency = 84e6

// ...

switch {
case localFrequency >= apb2Frequency / 2:
    conf = stm32.SPI_PCLK_2
case localFrequency >= apb2Frequency / 4:
    conf = stm32.SPI_PCLK_4
// ...
case localFrequency >= apb2Frequency / 128:
    conf = stm32.SPI_PCLK_128
default:
    conf = stm32.SPI_PCLK_256
}
```

This way the calculation is very explicit. This kind of calculation is always done at compile time, so you don't need to worry about any overhead it might introduce.
You can take a look here as an example: https://github.com/tinygo-org/tinygo/pull/1446